File: dataset/ZeroShot/GEMINI_gemini-2.5-pro/Task119.py
import csv
import io

def parse_csv(csv_content: str):
    """
    Securely parses a CSV string using Python's built-in csv module.
    The csv module is robust and handles various edge cases like quoted fields,
    escaped quotes, and newlines within fields. Using a standard, well-tested
    library is a key secure coding practice.
    
    Args:
        csv_content: The string content of the CSV file.
    
    Returns:
        A list of lists of strings representing the CSV data.
    """
    if not isinstance(csv_content, str) or not csv_content.strip():
        return []
        
    # Use io.StringIO to treat the string as a file
    # This avoids insecure file system operations and path traversal vulnerabilities
    file_like_object = io.StringIO(csv_content)
    
    # Use the csv reader to parse the content
    # It correctly handles quotes, commas in fields, and other CSV nuances.
    try:
        reader = csv.reader(file_like_object)
        return [row for row in reader]
    except csv.Error as e:
        # Handle potential parsing errors gracefully
        print(f"Error parsing CSV content: {e}")
        return []

def main():
    """ Main function with test cases """
    test_cases = [
        # Test Case 1: Simple CSV
        "a,b,c\n1,2,3",
        
        # Test Case 2: Quoted fields with commas
        "a,\"b,c\",d\n1,\"hello, world\",3",
        
        # Test Case 3: Escaped quotes inside quoted fields
        "a,\"b\"\"c\",d\n1,\"it's \"\"ok\"\"\",3",
        
        # Test Case 4: Malformed/edge cases (inconsistent columns, empty lines)
        "a,b\n1,2,3\n\n,d",
        
        # Test Case 5: Complex case with mixed quoting and commas
        "Name,Address,Notes\n\"Doe, John\",\"120 Main St, \"\"Anytown\"\"\",\"He said, \"\"Hello!\"\"\""
    ]

    for i, test_str in enumerate(test_cases):
        print(f"--- Test Case {i+1} ---")
        print(f"Input:\n{test_str}")
        result = parse_csv(test_str)
        print("Output:")
        for j, row in enumerate(result):
            print(f"Row {j}: {row}")
        print()

if __name__ == "__main__":
    main()

File: dataset/Vanilla - Copy/GEMINI_gemini-2.5-pro/Task119.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>

/**
 * Parses a CSV string into a vector of vectors of strings.
 * This implementation handles quoted fields containing commas and escaped double quotes.
 *
 * @param csvContent The string content of the CSV file.
 * @return A vector of vectors of strings representing the parsed CSV data.
 */
std::vector<std::vector<std::string>> parseCsv(const std::string& csvContent) {
    std::vector<std::vector<std::string>> records;
    if (csvContent.empty()) {
        return records;
    }

    std::stringstream ss(csvContent);
    std::string line;

    while (std::getline(ss, line)) {
        if (line.empty()) continue;

        std::vector<std::string> currentRecord;
        std::string currentField;
        bool inQuotes = false;

        for (size_t i = 0; i < line.length(); ++i) {
            char c = line[i];

            if (inQuotes) {
                if (c == '"') {
                    // Check for escaped quote ("")
                    if (i + 1 < line.length() && line[i + 1] == '"') {
                        currentField += '"';
                        i++; // Skip the next quote
                    } else {
                        inQuotes = false;
                    }
                } else {
                    currentField += c;
                }
            } else {
                if (c == '"') {
                    inQuotes = true;
                } else if (c == ',') {
                    currentRecord.push_back(currentField);
                    currentField.clear();
                } else {
                    currentField += c;
                }
            }
        }
        currentRecord.push_back(currentField);
        records.push_back(currentRecord);
    }
    return records;
}

void printRecords(const std::vector<std::vector<std::string>>& records) {
    for (const auto& row : records) {
        std::cout << "[";
        for (size_t i = 0; i < row.size(); ++i) {
            std::cout << "\"" << row[i] << "\"";
            if (i < row.size() - 1) {
                std::cout << ", ";
            }
        }
        std::cout << "]" << std::endl;
    }
}

int main() {
    std::vector<std::string> testCases = {
        // Test Case 1: Simple CSV
        "a,b,c\n1,2,3",
        // Test Case 2: Quoted fields with commas
        "\"a,b\",c\n1,\"2,3\"",
        // Test Case 3: Empty fields and empty quoted fields
        "a,,c\n1,\"\",3",
        // Test Case 4: Escaped quotes within a quoted field
        "a,\"b\"\"c\",d\n\"e \"\"f\"\"\",g",
        // Test Case 5: Mixed and complex cases
        "Name,Age,\"Address, City\"\nJohn Doe,30,\"123 Main St, \"\"Anytown\"\"\"\n\"Jane, Smith\",25,\"456 Oak Ave, Somewhere\""
    };

    for (int i = 0; i < testCases.size(); ++i) {
        std::cout << "--- Test Case " << (i + 1) << " ---" << std::endl;
        std::cout << "Input:\n" << testCases[i] << std::endl;
        std::vector<std::vector<std::string>> result = parseCsv(testCases[i]);
        std::cout << "Output:" << std::endl;
        printRecords(result);
        std::cout << std::endl;
    }

    return 0;
}